### DISCUSSION TITLE
How to generate a (bad) query string with equal sign in Mojo::URL

### DISCUSSION THREAD

**Discussion by fluca1978** [mojolicious/mojo#2157] [General]
I need to generate dynamically URLs with a query string like `table=XYZ=all` (note that the parameter is `table` and the value includes an equal sign).

```
my $url = Mojo::URL->new( q{https://localhost/db?table=XYZ=all} );
say $url->query;
```

this works and prints `table=XYZ=all` while the following one:

```
my $url = Mojo::URL->new( q{https://localhost/db} );
$url->query( table => 'XYZ=all' );
say $url->query;
```

prints `table=%3DXYZ%3Dall`, unluckily the backend cannot understand the (correctly) escaped equal sign.
Is there a way I can generate an unescaped value for the parameter?

**Comment by guest20** [comment]
If you just pass a single string to `query`: 


```perl
say Mojo::URL->new("https://localhost/db?previous=1")->query("thing=more=things")
```
will give you:

`https://localhost/db?thing=more=things`

**Comment by fluca1978** [comment]
> If you just pass a single string to `query`:
> 
> ```perl
> say Mojo::URL->new("https://localhost/db?previous=1")->query("thing=more=things")
> ```

Thanks, this is almost good. 
My problem is that I have multiple parameters to pass to the URL, that's why I would prefer to use `query` passing an hash, but if there's no way I have to build the query somehow (e.g., `sprintf`) and pass as a single string to `query`.

**Reply by guest20** [reply]
You could also sub-class `Mojo::Parameters` and adjust your sublcass's `to_string` method to do your special magic... You then use an instance of your sub-class as `query => ` when creating your `Mojo::URL` objects.

Be sure to write some tests for all this though, 'cause it can be very frustrating to maintain this kind of jank in your own project, especially when you forget it's in there and you're pulling your hair trying to figure out why the code doesn't work like the docs on cpan say...

**Reply by jberger** [reply]
At that point just make a new instance of Mojo::Parameters and join its results with what you have there.

```perl
my $params = Mojo::Parameters->new(foo => "bar", baz => "bat"); 
say Mojo::URL->new("https://localhost/db?previous=1")->query("thing=more=things&" . $params)
```

Its not pretty, and you'd probably want to check that you have some params before you join it with a trailing `&`, but there are all kinds of ways to do that

**Comment by fluca1978** [comment]
```
my $params = Mojo::Parameters->new(foo => "bar", baz => "bat"); 
say Mojo::URL->new("https://localhost/db?previous=1")->query("thing=more=things&" . $params)
```

the above sounds enough good to me. In my application I deal with a lot of URLs, and only this one has the problem, so I can deal with a "not so good looking code" for one case.

**Reply by guest20** [reply]
If you do anything with that object, Mojo::URL will normalise the params and your = will get escaped
